File: BitMasking_Problems/FlippingAnImage.js

// Problem link :- https://leetcode.com/problems/flipping-an-image/description/?envType=problem-list-v2&envId=bit-manipulation

let flipAndInvertImage = function(image) {
     // iterate through each row of the image matrix
    for(let row of image){
        // start pointer for the row (left side)
        let left = 0;

        // end pointer for the row (right side)
        let right = row.length - 1;

        // swap and invert simultaneously
        while(left <= right){
            // swap the left and right elements
            // XOR operation: 1 ^ 1 = 0, 0 ^ 1 = 1, so row[left] ^ 1 will invert the value
            let temp = row[left] ^ 1;
            row[left] = row[right] ^ 1;
            row[right] = temp;
            
            // move pointers towards the center
            left++;
            right--
        }
    }
    // return the modified image
    return image;
};

let image = [[1,1,0],[1,0,1],[0,0,0]];
console.log(flipAndInvertImage(image));

File: BitMasking_Problems/FlippingAnImage.py

# Problem link :- https://leetcode.com/problems/flipping-an-image/description/?envType=problem-list-v2&envId=bit-manipulation

from typing import List

def flipAndInvertImage(image: List[List[int]]) -> List[List[int]]:
    # iterate through each row of the matrix
    for row in image:
        # start and end pointers for the row
        left, right = 0, len(row) - 1  
        
        # swap and invert simultaneously
        while left <= right:
            # XOR operation to invert the values
            row[left], row[right] = row[right] ^ 1, row[left] ^ 1
            
            # move pointers towards the center
            left += 1
            right -= 1
    # return the modified image       
    return image  

image = [[1,1,0],[1,0,1],[0,0,0]]
print(flipAndInvertImage(image))